Remove duplicate function name in its aliases list

goldmedal opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #10658 

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?
yes
Also checked starting the CLI with the DEBUG level.
```
RUST_LOG=DEBUG ./target/debug/datafusion-cli
```
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?
No
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

goldmedal commented on datafusion/optimizer/src/push_down_limit.rs:
```diff
@@ -65,7 +65,6 @@ impl OptimizerRule for PushDownLimit {
         };
 
         let Limit { skip, fetch, input } = limit;
-        let input = input;
```
I'm not sure why the clippy check in CI didn't detect this but it's showed when I ran clippy in my local.

goldmedal commented on datafusion/core/src/execution/context/mod.rs:
```diff
@@ -2893,21 +2893,21 @@ mod tests {
             let udaf = new_state.register_udaf(function).unwrap();
             if let Some(udaf) = udaf {
-                assert!(false, "Function {} already registered", udaf.name());
+                unreachable!("Function {} already registered", udaf.name());
```
Clippy suggests me to use `panic` or `unreachable` instead of `assert`. I'm not sure which one is better.

alamb commented on datafusion/core/src/execution/context/mod.rs:
```diff
@@ -2860,6 +2863,57 @@ mod tests {
+            function_factory: None,
+        };
+
+        for function in all_default_functions() {
```
This is a nice test -- though it is basically a copy of SessionContext::new

I wonder if we could make a test that is a bit simpler for example by creating a new 
 https://docs.rs/datafusion/latest/datafusion/execution/registry/struct.MemoryFunctionRegistry.html

Something like (untested)
```rust
let registry = MemoryFunctionRegistry::new();
for function in all_default_array_functions() {
  let existing_function = new_state.register_udf(function);
  assert!(existing_function.is_none(), "{} was already registered", function.name()
}
// and similarly for the aggregate and array functins
```

🤔 
```

alamb commented:
Thank you @goldmedal  🙏

goldmedal commented:
Thanks, @alamb!

Actually, I tried to add some tests for `MemoryFunctionRegistry` before I fixed it, but I found everything was fine. I guess the reason is that we don't insert aliases when using `MemoryFunctionRegistry#register_udf`.

https://github.com/apache/datafusion/blob/4709fc65f7debc143696fa3a23ab6569ec8a383c/datafusion/execution/src/registry.rs#L174-L176

Then, I found that the aliases are only inserted by `SessionState` in

https://github.com/apache/datafusion/blob/4709fc65f7debc143696fa3a23ab6569ec8a383c/datafusion/core/src/execution/context/mod.rs#L2278-L2283

That's why I added tests for `SessionState`.

I'm not sure, but maybe we should also insert aliases in `MemoryFunctionRegistry`. What do you think?

alamb commented:
> I'm not sure, but maybe we should also insert aliases in MemoryFunctionRegistry. What do you think?

I think this would make sense -- though maybe we can do it as a follow on PR

alamb commented on datafusion/core/src/execution/context/mod.rs:
```diff
@@ -2860,6 +2863,57 @@ mod tests {
 
+    #[tokio::test]
+    async fn test_register_default_functions() -> Result<()> {
+        let config = SessionConfig::new();
```
While I understand the proposal to test with code like `SessionState::new` , I think it is problematic because if SessionState::new is changed, then this test will no longer match what it is doing.

I think the core of this test is verifying that there are no duplicate names in the alias lists.

Thus, perhaps we could write tests for `all_default_functions`, etc like

```rust
let mut names = HashSet::new();
for func in all_default_functions() {
  assert(names.insert(func.name()).is_none(), "func.name duplicated")
  for alias in func.aliases() {
    assert(names.insert(alias).is_none(), "alias duplicated")
   }
}
```

What do you think?

goldmedal replied:
Good idea. I'll add this test and remove the test for SessionState::new. Thanks

alamb reviewed (commented):
Thank you for this find @goldmedal  -- very cool. I think we can simplify the test a bit which would be good to consider.

goldmedal commented:
I'm confused about the new `median` function. I'm not sure why the name is case-sensitive.

```
Planning query: Plan("Invalid function 'median'.\nDid you mean 'MEDIAN'?\n\n
```

I removed the duplicate name as I did for other functions in 4086688. However, the planner can't resolve the lowercase function name. I have tried and confirmed that other functions (e.g., `first_value` and `approx_median`) are case-insensitive. Only `median` is case-sensitive.

Do you have any idea? @alamb 

I found this function was created in #10644.
Gentle ping @jayzhan211 , do you have any idea about it?

goldmedal commented:
I think I got it now. In #10644, `MEDIAN` was removed from the AggregationFunction list and reimplemented as a UDAF. It looks like the list of aggregation functions handles some name mapping for case insensitivity.

See:
https://github.com/apache/datafusion/blob/52c4f3cce68afc45f25599b3917e06f0c6b63192/datafusion/expr/src/aggregate_function.rs#L117
https://github.com/apache/datafusion/blob/52c4f3cce68afc45f25599b3917e06f0c6b63192/datafusion/expr/src/aggregate_function.rs#L171

That's why we don't need to add a lowercase alias for `FIRST_VALUE` in `first_last.rs`. It seems `FIRST_VALUE` has been redesigned as a UDAF but hasn't been removed from the AggregationFunction list. I'm not sure if this is an issue.

I reverted the change for `median` and removed the name testing for case insensitivity. However, I think adding a lowercase alias for `MEDIAN` is a workaround to handle the function name case insensitivity issue for UDAF. Maybe we need a formal way to handle this, but I'm not sure.

jayzhan211 commented:
> That's why we don't need to add a lowercase alias for `FIRST_VALUE` in `first_last.rs`. It seems `FIRST_VALUE` has been redesigned as a UDAF but hasn't been removed from the AggregationFunction list. I'm not sure if this is an issue.

`first_value` is not added in alias because the function is not fully converted to UDAF yet, so it will fail some tests. #10648 has alias for lower case

so as `approx_median` since it is not UDAF yet.

> I think adding a lowercase alias for MEDIAN is a workaround to handle the function name case insensitivity issue for UDAF
Why adding a lowercase alias is a workaround for UDAF? Without alias, we will need to check the lowercase in many places.

goldmedal commented:
> `first_value` is not added in alias because the function is not fully converted to UDAF yet, so it will fail some tests. #10648 has alias for lower case
> 
> so as `approx_median` since it is not UDAF yet.
> 

Thanks for the information! @jayzhan211 

> > I think adding a lowercase alias for MEDIAN is a workaround to handle the function name case insensitivity issue for UDAF
> 
> Why adding a lowercase alias is a workaround for UDAF? Without alias, we will need to check the lowercase in many places.

Since I thought we needed to list all case patterns for the function name (e.g., `Median`, `MEDian`), I called it a workaround. However, I found I was wrong. After some testing, I discovered something interesting. For DataFusion, if we query a function
```
SELECT Median(xx) FROM tt
```
The function call will be transferred to `median`. Even query it use
```
SELECT MEDIAN(xx) FROM tt
```
It will also be changed to `median`. The uppercase name is only be matched when invoke this function with qutoes
```
SELECT "MEDIAN"(xx) FROM tt
```
So, how about setting UDAF names to lowercase (e.g., `median`, `first_value`)? This way, we don't need another uppercase alias. What do you think?

alamb reviewed (approved):
Tahnk you @goldmedal  and @jayzhan211  -- I think this is a really nice improvement. Not only does it fix the bug it also makes the code easier to work with and avoids introducing the same regression again

I took the liberty of merging up from main to resolve a merge conflict

alamb commented:
> So, how about setting UDAF names to lowercase (e.g., median, first_value)? This way, we don't need another uppercase alias. What do you think?

This makes sense to me 

Shall we file a follow on ticket for it? I don't think we need to fix it in this PR (as this PR doesn't make the situation any better or worse, from what I can tell).

goldmedal commented:
> Shall we file a follow on ticket for it? I don't think we need to fix it in this PR (as this PR doesn't make the situation any better or worse, from what I can tell).

Thanks @alamb, I aggress with you. Fix it in a follow-up ticket is better.

jayzhan211 commented:
Thanks, @goldmedal and @alam. I think we can merge this

goldmedal commented:
Thanks again @alamb @jayzhan211

alamb commented:
> > Shall we file a follow on ticket for it? I don't think we need to fix it in this PR (as this PR doesn't make the situation any better or worse, from what I can tell).
> 
> Thanks @alamb, I agree with you. Fixing it in a follow-up ticket is better.

Filed https://github.com/apache/datafusion/issues/10695 to track
